File: app/src/main/java/com/criptext/uisample/SensorHandler.java
package com.criptext.uisample;

import android.app.Activity;
import android.content.Context;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.media.AudioManager;
import android.view.WindowManager;

import com.criptext.monkeykitui.recycler.audio.AudioPlaybackHandler;
import com.criptext.monkeykitui.recycler.audio.VoiceNotePlayer;

/**
 * Created by daniel on 5/12/16.
 */

public class SensorHandler implements SensorEventListener {

    private VoiceNotePlayer voiceNotePlayer;
    private boolean isProximityOn=false;
    private SensorManager mSensorManager;
    private Sensor mSensor;
    private AudioManager mAudioManager;
    private Activity activity;
    private float prevBrightness;
    private WindowManager.LayoutParams layout;

    public SensorHandler(VoiceNotePlayer voiceNotePlayer, Activity activity) {

        this.voiceNotePlayer = voiceNotePlayer;
        this.activity = activity;
        layout = activity.getWindow().getAttributes();
        mAudioManager = (AudioManager) activity.getSystemService(Context.AUDIO_SERVICE);
        mSensorManager = (SensorManager) activity.getSystemService(Context.SENSOR_SERVICE);
        mSensor = mSensorManager.getDefaultSensor(Sensor.TYPE_PROXIMITY);
        mSensorManager.registerListener(this, mSensor, SensorManager.SENSOR_DELAY_FASTEST);

    }

    @Override
    public void onSensorChanged(SensorEvent event) {

        if (event.values[0] < mSensor.getMaximumRange()) {

            if(voiceNotePlayer !=null && voiceNotePlayer.isPlayingAudio()) {
                voiceNotePlayer.onPauseButtonClicked();
                voiceNotePlayer.releasePlayer();
                voiceNotePlayer.initPlayerWithFrontSpeaker();
                activity.setVolumeControlStream(AudioManager.STREAM_VOICE_CALL);
                isProximityOn=true;
                prevBrightness = layout.screenBrightness;
                layout.screenBrightness = 0;
                activity.getWindow().setAttributes(layout);
            }

        } else {

            if(voiceNotePlayer !=null && isProximityOn){
                voiceNotePlayer.onPauseButtonClicked();
                voiceNotePlayer.releasePlayer();
                voiceNotePlayer.initPlayer();
                mAudioManager.setMode(AudioManager.MODE_NORMAL);
                isProximityOn=false;
                layout.screenBrightness = prevBrightness;
                activity.getWindow().setAttributes(layout);
            }
        }
    }

    @Override
    public void onAccuracyChanged(Sensor sensor, int accuracy) {

    }

    public void onPause(){
        mAudioManager.setMode(AudioManager.MODE_NORMAL);
    }

    public void onStop(){
        if(isProximityOn){
            mAudioManager.setMode(AudioManager.MODE_NORMAL);
            isProximityOn=false;
            layout.screenBrightness = prevBrightness;
            activity.getWindow().setAttributes(layout);
        }
    }

    public void onDestroy(){
        mSensorManager.unregisterListener(this);
    }

}

File: monkeykitui/src/main/kotlin/com/criptext/monkeykitui/recycler/audio/SensorHandler.kt
package com.criptext.monkeykitui.recycler.audio

import android.app.Activity
import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.media.AudioManager
import android.os.PowerManager
import android.util.Log


/**
 * Class that uses a Proximity Sensor to turn off the screen when the user has the device close to
 * face. If audio is playing, changes the audio stream to voice call mode.
 * Created by Gabriel on 5/13/16.
 */
class SensorHandler(private val voiceNotePlayer: VoiceNotePlayer?, private val ctx: Context) : SensorEventListener {
    var isProximityOn = false
    private set
    private val mSensorManager: SensorManager
    private val mSensor: Sensor
    private val mAudioManager: AudioManager
    private val mProximityWakeLock: PowerManager.WakeLock? by lazy {
        val pm = ctx.getSystemService(Context.POWER_SERVICE) as PowerManager
        var field = 0x00000020
        try {
            if(android.os.Build.VERSION.SDK_INT >= 21)
                field = PowerManager.PROXIMITY_SCREEN_OFF_WAKE_LOCK
            else
                // Yeah, this is hidden field in older versions.
                field = PowerManager::class.java.javaClass.getField("PROXIMITY_SCREEN_OFF_WAKE_LOCK").getInt(null)

        } catch (ignored: Throwable) { /* Undocumented field in this version */ }

        pm.newWakeLock(field, "ScreenOffLock")
    }

    init {
        mAudioManager = ctx.getSystemService(Context.AUDIO_SERVICE) as AudioManager
        mSensorManager = ctx.getSystemService(Context.SENSOR_SERVICE) as SensorManager
        mSensor = mSensorManager.getDefaultSensor(Sensor.TYPE_PROXIMITY)
        mSensorManager.registerListener(this, mSensor, SensorManager.SENSOR_DELAY_FASTEST)

    }

    override fun onSensorChanged(event: SensorEvent) {

        if (event.values[0] < mSensor.maximumRange) {

            if (voiceNotePlayer != null && voiceNotePlayer.isPlayingAudio) {
                voiceNotePlayer.onPauseButtonClicked()
                voiceNotePlayer.releasePlayer()
                voiceNotePlayer.initPlayerWithFrontSpeaker()
                //ctx.volumeControlStream = AudioManager.STREAM_VOICE_CALL
                mAudioManager.mode = AudioManager.STREAM_VOICE_CALL
                isProximityOn = true
                if(!(mProximityWakeLock?.isHeld ?: false)) mProximityWakeLock?.acquire()
            }

        } else {

            if (voiceNotePlayer != null && isProximityOn) {
                voiceNotePlayer.onPauseButtonClicked()
                voiceNotePlayer.releasePlayer()
                voiceNotePlayer.initPlayer()
                mAudioManager.mode = AudioManager.MODE_NORMAL
                isProximityOn = false
                if((mProximityWakeLock?.isHeld ?: false)) mProximityWakeLock?.release()
                if(voiceNotePlayer.currentlyPlayingItem != null)
                    voiceNotePlayer.uiUpdater?.rebindAudioHolder(voiceNotePlayer.currentlyPlayingItem!!.item)
            }
        }
    }

    override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {

    }

    fun onDestroy() {
        mSensorManager.unregisterListener(this)
    }
}
